File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/SubjectConfigurationUpdate.java
package com.snowplowanalytics.snowplow.internal.tracker;

import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.configuration.SubjectConfiguration;
import com.snowplowanalytics.snowplow.util.Size;

public class SubjectConfigurationUpdate extends SubjectConfiguration {

    @Nullable
    public SubjectConfiguration sourceConfig;

    // userId flag

    public boolean userIdUpdated;

    @Nullable
    public String getUserId() {
        return (sourceConfig == null || userIdUpdated) ? super.userId : sourceConfig.userId;
    }

    // networkUserId flag

    public boolean networkUserIdUpdated;

    @Nullable
    public String getNetworkUserId() {
        return (sourceConfig == null || networkUserIdUpdated) ? super.networkUserId : sourceConfig.networkUserId;
    }

    // domainUserId flag

    public boolean domainUserIdUpdated;

    @Nullable
    public String getDomainUserId() {
        return (sourceConfig == null || domainUserIdUpdated) ? super.domainUserId : sourceConfig.domainUserId;
    }

    // useragent flag

    public boolean useragentUpdated;

    @Nullable
    public String getUseragent() {
        return (sourceConfig == null || useragentUpdated) ? super.useragent : sourceConfig.useragent;
    }

    // ipAddress flag

    public boolean ipAddressUpdated;

    @Nullable
    public String getIpAddress() {
        return (sourceConfig == null || ipAddressUpdated) ? super.ipAddress : sourceConfig.ipAddress;
    }

    // timezone flag

    public boolean timezoneUpdated;

    @Nullable
    public String getTimezone() {
        return (sourceConfig == null || timezoneUpdated) ? super.timezone : sourceConfig.timezone;
    }

    // language flag

    public boolean languageUpdated;

    @Nullable
    public String getLanguage() {
        return (sourceConfig == null || languageUpdated) ? super.language : sourceConfig.language;
    }

    // screenResolution flag

    public boolean screenResolutionUpdated;

    @Nullable
    public Size getScreenResolution() {
        return (sourceConfig == null || screenResolutionUpdated) ? super.screenResolution : sourceConfig.screenResolution;
    }

    // screenViewPort flag

    public boolean screenViewPortUpdated;

    @Nullable
    public Size getScreenViewPort() {
        return (sourceConfig == null || screenViewPortUpdated) ? super.screenViewPort : sourceConfig.screenViewPort;
    }

    // colorDepth flag

    public boolean colorDepthUpdated;

    @Nullable
    public Integer getColorDepth() {
        return (sourceConfig == null || colorDepthUpdated) ? super.colorDepth : sourceConfig.colorDepth;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/SubjectConfigurationUpdate.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.core.tracker

import com.snowplowanalytics.snowplow.configuration.SubjectConfiguration
import com.snowplowanalytics.snowplow.util.Size

class SubjectConfigurationUpdate : SubjectConfiguration() {
    var sourceConfig: SubjectConfiguration? = null
    private var userIdUpdated = false
    private var networkUserIdUpdated = false
    private var domainUserIdUpdated = false
    private var useragentUpdated = false
    private var ipAddressUpdated = false
    private var timezoneUpdated = false
    private var languageUpdated = false
    private var screenResolutionUpdated = false
    private var screenViewPortUpdated = false
    private var colorDepthUpdated = false

    override var userId: String?
        get() = if (sourceConfig == null || userIdUpdated) super.userId else sourceConfig!!.userId
        set(value) {
            super.userId = value
            userIdUpdated = true
        }
    
    override var networkUserId: String?
        get() = if (sourceConfig == null || networkUserIdUpdated) super.networkUserId else sourceConfig!!.networkUserId
        set(value) {
            super.networkUserId = value
            networkUserIdUpdated = true
        }
    
    override var domainUserId: String?
        get() = if (sourceConfig == null || domainUserIdUpdated) super.domainUserId else sourceConfig!!.domainUserId
        set(value) {
            super.domainUserId = value
            domainUserIdUpdated = true
        }
    
    override var useragent: String?
        get() = if (sourceConfig == null || useragentUpdated) super.useragent else sourceConfig!!.useragent
        set(value) {
            super.useragent = value
            useragentUpdated = true
        }
    
    override var ipAddress: String?
        get() = if (sourceConfig == null || ipAddressUpdated) super.ipAddress else sourceConfig!!.ipAddress
        set(value) {
            super.ipAddress = value
            ipAddressUpdated = true
        }
    
    override var timezone: String?
        get() = if (sourceConfig == null || timezoneUpdated) super.timezone else sourceConfig!!.timezone
        set(value) {
            super.timezone = value
            timezoneUpdated = true
        }
    
    override var language: String?
        get() = if (sourceConfig == null || languageUpdated) super.language else sourceConfig!!.language
        set(value) {
            super.language = value
            languageUpdated = true
        }
    
    override var screenResolution: Size?
        get() = if (sourceConfig == null || screenResolutionUpdated) super.screenResolution else sourceConfig!!.screenResolution
        set(value) {
            super.screenResolution = value
            screenResolutionUpdated = true
        }
    
    override var screenViewPort: Size?
        get() = if (sourceConfig == null || screenViewPortUpdated) super.screenViewPort else sourceConfig!!.screenViewPort
        set(value) {
            super.screenViewPort = value
            screenViewPortUpdated = true
        }
    
    override var colorDepth: Int?
        get() = if (sourceConfig == null || colorDepthUpdated) super.colorDepth else sourceConfig!!.colorDepth
        set(value) {
            super.colorDepth = value
            colorDepthUpdated = true
        }
}
